### PR TITLE
Ensure URI creation is cross platform compatible

### PR DESCRIPTION
**Pull request by anamnavi**
# PR Summary

Fixed URI creation in Register-PSResource and Set-PSResource cmdlets. The user input for URL is now of type String instead of URL and in ProcessRecord we convert it to type URI while ensuring it's compatible cross platform and with absolute paths (and if it's detected to not be an absolute path, resolve it to the relative path)

## PR Context

Before, this was breaking tests in Linux.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [ ] None
    - **OR**
    - [ ] [Documentation needed]()
        - [ ] Issue filed: 
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [ ] [Documentation needed]()
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [ ] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary.

### CODE REVIEW

**Review by alerickson — CHANGES_REQUESTED** [review]
We discussed with Paul offline, but the only change I'd recommend is pulling out the TryCreateURL method into the Utils class and just passing in the PSCmdlet as a param so that you can access the SessionState.

Everything else looks good to me

File: src/code/RegisterPSResourceRepository.cs [hunk 1]
```diff
@@ -156,9 +139,15 @@ protected override void ProcessRecord()
             switch (ParameterSetName)
             {
                 case NameParameterSet:
+                    bool isUrlValid = Utils.TryCreateValidUrl(URL, this, out _url, out ErrorRecord errorRecord);
```
Comment (PaulHigin) [inline src/code/RegisterPSResourceRepository.cs]: Nitpick: 
```csharp
Utils.TryCreateValidUrl(URL, cmdletPassedIn: this, out _url, out ErrorRecord errorRecord);
```

File: src/code/RegisterPSResourceRepository.cs [hunk 2]
```diff
@@ -334,13 +323,11 @@ private PSRepositoryInfo RepoValidationHelper(Hashtable repo)
                 return null;
             }
 
-            if (!Uri.TryCreate(repo["URL"].ToString(), UriKind.Absolute, out Uri repoURL))
+            string url = repo["Url"].ToString();
+            bool isUrlValid = Utils.TryCreateValidUrl(url, this, out Uri repoURL, out ErrorRecord errorRecord);
+            if (!isUrlValid)
```
Comment (PaulHigin) [inline src/code/RegisterPSResourceRepository.cs]: Nitpick:
``` csharp
if (!Utils.TryCreateValidUrl(
    urlString:  repo["Url"].ToString(),
    cmdlet: this,
    out Uri repoURL,
    out ErrorRecord errorRecord))
{

}
```

File: src/code/SetPSResourceRepository.cs [hunk 3]
```diff
@@ -129,14 +112,23 @@ protected override void BeginProcessing()
 
         protected override void ProcessRecord()
         {
+            if (!String.IsNullOrEmpty(URL))
```
Comment (PaulHigin) [inline src/code/SetPSResourceRepository.cs]: Suggest:
```csharp
if (MyInvocation.BoundParameters.ContainsKey(nameof(URL)))
{

}
```

File: src/code/SetPSResourceRepository.cs [hunk 4]
```diff
@@ -243,15 +235,27 @@ private List<PSRepositoryInfo> RepositoriesParameterSetHelper()
...
+                            new PSInvalidOperationException("Repository url cannot be null if provided"),
+                            "NullURLForRepositoriesParameterSetUpdate",
+                            ErrorCategory.InvalidArgument,
+                            this));
+                    return null;
+                }
+
+                string url = repo["Url"].ToString();
+                bool isUrlValid = Utils.TryCreateValidUrl(url, this, out repoURL, out ErrorRecord errorRecord);
+                if (!isUrlValid)
```
Comment (PaulHigin) [inline src/code/SetPSResourceRepository.cs]: Same nitpick comment as above.

File: src/code/Utils.cs [hunk 5]
```diff
@@ -171,6 +172,55 @@ public static string[] GetStringArray(ArrayList list)
...
+            if (!urlString.StartsWith(Uri.UriSchemeHttps) &&
+                !urlString.StartsWith(Uri.UriSchemeHttp) &&
+                !urlString.StartsWith(Uri.UriSchemeFtp))
+            {
+                // url string could be of type (potentially) UriSchemeFile or invalid type
+                // can't check for UriSchemeFile because relative paths don't qualify as UriSchemeFile
+                try
+                {
+                    // this is needed for a relative path urlstring. Does not throw error for an absolute path
+                    string resolvedUrlString = cmdletPassedIn.SessionState.Path.GetResolvedPSPathFromPSPath(urlString)[0].Path;
```
Comment (PaulHigin) [inline src/code/Utils.cs]: Don't really need 'resolvedUrlString' variable since it is used only once.

**Review by PaulHigin — APPROVED** [review]
Just some style comments.

**Review by alerickson — APPROVED** [review]

**Review by PaulHigin — APPROVED** [review]

**Pull request merged**